apple/Sources/Sargon/Util/SargonObjectCodable.swift

```diff
@@ -0,0 +1,23 @@
+import Foundation
+import AnyCodable
+
+public protocol SargonObjectCodable: Codable {
+	init(jsonData: some DataProtocol) throws
+	func jsonData() -> Data
+}
+
+extension SargonObjectCodable {
```

Comment:
@iOS devs: can we do better than this? I think we MUST have AnyCodable or similar package, check commit history... I also managed to use Rob Napiers RNJSON package. It was slightly better actually but less known package.

Reply:
Some context: we don't use Codable on our 100 types in Sargon since Mozilla UniFFI does not translate Rust Serde (Serde is Rusts de facto standard JSON tool) to Codable for bindgen generated Swift types.

In a few rare cases we must be able to JSON code some of Sargons types, eg DeviceInfo, at least for first ingestion.

So the API is:
rust Sargon exposes toJSONData() on eg DeviceInfo and a constructor of DeviceInfo fromJSONData. 

We need to hook up those methods to Swifts Codable system...

Reply:
Why do we need the protocol at all, can't we just implement Codable directly?

Reply:
Ah alright, so you use AnyCodable as a bridge, to encode, you first serialise into data, then you deserialise into a dictionary and serialise _that_, but this time using the provided encoder. You can't ask Rust to just turn it into a dictionary directly?

Reply:
now you got it :D 

No, I cannot return `[String: AnyCodable]` from Rust...

Reply:
The purpose of the Protocol - is of course - so that I do not need to implement this `N` times, but just once, and get it for free for `N` many types, simply by `extension YetAnotherType: SargonObjectCodable`

Reply:
as I said on slack, what I wanted input on was, can we do it even "better":

```swift
	public func encode(to encoder: any Encoder) throws {
		let dict = try JSONSerialization.jsonObject(with: self.jsonData(), options: []) as! NSDictionary
		let anyEncodable = AnyEncodable(dict)
		var container = encoder.singleValueContainer()
		try container.encode(anyEncodable)
	}
	
	public init(from decoder: any Decoder) throws {
		let container = try decoder.singleValueContainer()
		let anyDecodable = try container.decode(AnyDecodable.self)
		let jsonData = try JSONSerialization.data(withJSONObject: anyDecodable.value)
		try self.init(jsonData: jsonData)
	}
```

I dont think we can do it much better... and I think we must have a corresponsing `AnyCodable` or `JSONValue` enum...

Reply:
>  I do not need to implement this N times, but just once
That's what I meant by repeated boilerplate. Seems like my suggestion should work. I mean you already have repeated boilerplate for each time, where you implement your protocol. That's where I propose that you implement Codable directly, and put all the boilerplate in two free wrapper functions, one for encoding and one for decoding.

You could alternatively make sure the Rust methods have the same name for all types, then you could keep your protocol, but you could probably avoid those bridging methods.

Reply:
Sounds like you now are talking about the `func jsonData() -> Data` part? That cannot change. Unfortunately these are not Rust methods, those are Rust **functions** (global scope), so MUST have a unique name. So we always need to bridge global scope functions -> Swift methods/inits.

Reply:
Correct, my alternative suggestion doesn't work if all you have from Rust is global methods with different names.